eggbox_potential_sampler/model_based_mco/mco.py

```diff
@@ -0,0 +1,132 @@
+import subprocess
+import sys
+import logging
+
+from skopt import Optimizer
+
+from traits.api import (
+    HasStrictTraits, List, Str, Instance, Interface, provides
+)
+
+from force_bdss.api import (
+    BaseMCO, execute_workflow, Workflow, DataValue
+)
+
+
+log = logging.getLogger(__name__)
+
+
+class ModelBasedOptimizationMCO(BaseMCO):
+    """ This MCO draws random samples in [0, 1] and samples a random
+    eggbox potential constructed by EggboxPESDataSource. Calls to
+    force_bdss can be made either as a separate subprocess, or internally.
+
+    """
+    def run(self, model):
+        kpis = model.kpis
+        application = self.factory.plugin.application
+        if model.evaluation_mode == "Subprocess":
+            single_point_evaluator = SubprocessSinglePointEvaluator(
+                sys.argv[0], application.workflow_filepath
+            )
+        else:
+            single_point_evaluator = InternalSinglePointEvaluator(
+                application.workflow,
+                model.parameters
+            )
+
+        optimizer = Optimizer([(0.0, 1.0)
+                               for i in range(len(model.parameters))],
+                              base_estimator=model.estimator,
+                              n_initial_points=model.num_random_trials)
+
+        n_total = model.num_trials + model.num_random_trials
+        counter = 0
+        fit = False
+        while counter < n_total:
+            counter += 1
+            if counter >= model.num_random_trials:
+                fit = True
+            log.info("MCO iteration {}/{}".format(counter, n_total))
+            trial_position = optimizer.ask()
+
+            kpis = single_point_evaluator.evaluate(trial_position)
+            # use first KPI only for now
+            optimizer.tell(trial_position, kpis[0], fit=fit)
```

Comment:
As mentioned up top, this should be changed to a weighted sum of KPIs, depending on the user input, but this currently isn't passed to the plugins.